crates/client/exec/src/block_context.rs

```
@@ -36,10 +49,10 @@ impl<'a> ExecutionContext<'a> {
             }
         };
 
-        CachedState::new(BlockifierStateAdapter::new(self.backend, on_top_of), GlobalContractCache::new(16))
+        CachedState::new(BlockifierStateAdapter::new(Arc::clone(&self.backend), on_top_of))
     }
 
-    pub fn new(backend: &'a DeoxysBackend, block_info: &DeoxysMaybePendingBlockInfo) -> Result<Self, Error> {
+    pub fn new(backend: Arc<DeoxysBackend>, block_info: &DeoxysMaybePendingBlockInfo) -> Result<Self, Error> {
         let (db_id, protocol_version, block_number, block_timestamp, sequencer_address, l1_gas_price, l1_da_mode) =
             match block_info {
                 DeoxysMaybePendingBlockInfo::Pending(block) => (
```

in case of pending `block_number`, should we throw an error instead of return default `0`

---

So the thinking here is: This part was made to work when the actual genesis block is in a pending state (meaning, we are executing on a pending block and there are no latest block in db)
This is the reason this unwrap_or(0) exists: in this case we would be executing genesis

Tracing the execution using the trace rpcs require reexecuting old blocks, so this is useful for them